File: raincoat-fabric/src/main/java/dev/uten2c/raincoat/updater/Updater.java
package dev.uten2c.raincoat.updater;

import dev.uten2c.raincoat.RaincoatMod;
import dev.uten2c.raincoat.util.UpdaterUtils;
import net.fabricmc.loader.api.FabricLoader;
import net.fabricmc.loader.api.entrypoint.PreLaunchEntrypoint;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.xml.parsers.DocumentBuilderFactory;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.Files;

public class Updater implements PreLaunchEntrypoint {
    private static @Nullable String updatableVersion;

    @Override
    public void onPreLaunch() {
        asyncFetchReleaseVersion();
    }

    public static boolean isUpdateAvailable() {
        return Updater.updatableVersion != null;
    }

    @SuppressWarnings("OptionalGetWithoutIsPresent")
    public static void tryStartDownload(@NotNull Runnable completeCallback, @NotNull Runnable errorCallback) {
        var updatableVersion = Updater.updatableVersion;
        if (updatableVersion == null) {
            errorCallback.run();
            return;
        }
        try (var in = downloadMod(updatableVersion)) {
            var modContainer = FabricLoader.getInstance().getModContainer(RaincoatMod.MOD_ID).get();
            var jarPath = modContainer.getOrigin().getPaths().get(0);
            var out = new DataOutputStream(Files.newOutputStream(jarPath));
            out.write(in.readAllBytes());
            completeCallback.run();
        } catch (Exception e) {
            e.printStackTrace();
            errorCallback.run();
        }
    }

    @SuppressWarnings("OptionalGetWithoutIsPresent")
    public static void asyncFetchReleaseVersion() {
        new Thread(() -> {
            try {
                var releaseVersion = fetchReleaseVersion();
                if (releaseVersion == null) {
                    return;
                }
                var modContainer = FabricLoader.getInstance().getModContainer(RaincoatMod.MOD_ID).get();
                var currentVersion = modContainer.getMetadata().getVersion();
                if (!currentVersion.getFriendlyString().equals(releaseVersion)) {
                    updatableVersion = releaseVersion;
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }).start();
    }

    private static @Nullable String fetchReleaseVersion() throws IOException {
        var url = new URL(UpdaterUtils.MAVEN_METAFILE_URL);
        try (var in = download(url)) {
            var factory = DocumentBuilderFactory.newInstance();
            var builder = factory.newDocumentBuilder();
            var document = builder.parse(in);
            var element = document.getDocumentElement();
            var release = element.getElementsByTagName("release");
            return release.item(0).getTextContent();
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    private static InputStream downloadMod(@NotNull String version) throws Exception {
        var url = UpdaterUtils.getArtifactUrl(version);
        return download(url);
    }

    private static InputStream download(@NotNull URL url) throws Exception {
        var conn = (HttpURLConnection) url.openConnection();
        conn.setAllowUserInteraction(false);
        conn.setInstanceFollowRedirects(true);
        conn.setRequestMethod("GET");
        conn.connect();
        var responseCode = conn.getResponseCode();
        if (responseCode != HttpURLConnection.HTTP_OK) {
            throw new Exception("Http status code: " + responseCode);
        }
        return conn.getInputStream();
    }
}

File: raincoat-fabric/src/main/kotlin/dev/uten2c/raincoat/updater/Updater.kt
package dev.uten2c.raincoat.updater

import com.github.kittinunf.fuel.Fuel
import com.github.kittinunf.fuel.coroutines.awaitByteArray
import com.mojang.logging.LogUtils
import dev.uten2c.raincoat.MOD_ID
import dev.uten2c.raincoat.util.UpdaterUtils
import kotlinx.coroutines.*
import net.fabricmc.loader.api.FabricLoader
import net.fabricmc.loader.api.SemanticVersion
import java.nio.file.AccessDeniedException
import java.nio.file.Files
import javax.xml.parsers.DocumentBuilderFactory
import kotlin.jvm.optionals.getOrNull

object Updater {
    private val logger = LogUtils.getLogger()
    private val modContainer = FabricLoader.getInstance().getModContainer(MOD_ID).get()
    private var updatableVersion: String? = null
    val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    @JvmStatic
    val isUpdateAvailable: Boolean
        get() = updatableVersion != null

    suspend fun startDownload(): UpdateStatus {
        val updatableVersion = updatableVersion ?: return UpdateStatus.ErrorReturnToTitle("アップデート可能なバージョンが見つかりませんでした")
        val bytes = runCatching {
            Fuel.get(UpdaterUtils.getArtifactUrl(updatableVersion)).awaitByteArray()
        }.onFailure {
            logger.error("MODのダウンロードに失敗しました", it)
            return UpdateStatus.ErrorReturnToTitle("MODのダウンロードに失敗しました")
        }.getOrThrow()
        runCatching {
            val jarPath = modContainer.origin.paths[0]
            withContext(Dispatchers.IO) {
                Files.newOutputStream(jarPath).use { out ->
                    out.write(bytes)
                }
            }
        }.onFailure {
            if (it is AccessDeniedException) {
                logger.error("MODファイルへの書き込みに失敗しました", it)
                return UpdateStatus.ErrorReturnToTitle("MODファイルへの書き込みに失敗しました")
            }
            logger.error("予期しないエラーが発生しました", it)
            return UpdateStatus.ErrorShouldRestart("予期しないエラーが発生しました")
        }
        return UpdateStatus.Completed
    }

    @JvmStatic
    fun asyncCheckUpdate() {
        scope.launch {
            checkUpdate()
        }
    }

    private suspend fun checkUpdate() {
        fetchReleaseVersion()
            .onSuccess { releaseVersion ->
                runCatching {
                    SemanticVersion.parse(modContainer.metadata.version.friendlyString)
                }.onSuccess { currentVersion ->
                    if (shouldUpdate(currentVersion, releaseVersion)) {
                        updatableVersion = releaseVersion.friendlyString
                    }
                }.onFailure {
                    logger.warn("バージョンのパースに失敗しました")
                }
            }
            .onFailure {
                logger.error("最新バージョンの確認に失敗しました", it)
            }
    }

    private suspend fun fetchReleaseVersion(): Result<SemanticVersion> {
        return kotlin.runCatching {
            val bytes = Fuel.get(UpdaterUtils.MAVEN_METAFILE_URL).awaitByteArray()
            val factory = DocumentBuilderFactory.newInstance()
            val builder = factory.newDocumentBuilder()
            val document = builder.parse(bytes.inputStream())
            val element = document.documentElement
            val release = element.getElementsByTagName("release")
            val versionString = release.item(0).textContent
            SemanticVersion.parse(versionString)
        }
    }

    private fun shouldUpdate(current: SemanticVersion, latest: SemanticVersion): Boolean {
        return equalsMinecraftVersion(current, latest) &&
                current.buildKey.getOrNull() != latest.buildKey.getOrNull()
    }

    private fun equalsMinecraftVersion(a: SemanticVersion, b: SemanticVersion): Boolean {
        return a.getVersionComponent(0) == b.getVersionComponent(0) &&
                a.getVersionComponent(1) == b.getVersionComponent(1) &&
                a.getVersionComponent(2) == b.getVersionComponent(2)
    }
}
